### ISSUE TITLE
JSC_ILLEGAL_PROPERTY_ACCESS on Map and Set

### ISSUE THREAD

**Issue by nazar-pc** [google/closure-compiler#2677]
```javascript
var m = new Map;
m.set('x', 'x');
m['delete']('x');
var s = new Set;
s.add('x');
s['delete']('x');
```

```
nazar-pc@nazar-pc ~/n/google-closure-compiler> java -jar compiler.jar --version
Closure Compiler (http://github.com/google/closure-compiler)
Version: v20170910
Built on: 2017-09-12 14:27
nazar-pc@nazar-pc ~/n/google-closure-compiler> java -jar compiler.jar --compilation_level ADVANCED_OPTIMIZATIONS --language_in ES5 x.js
x.js:3: WARNING - Cannot do '[]' access on a struct
m['delete']('x');
  ^^^^^^^^

x.js:6: WARNING - Cannot do '[]' access on a struct
s['delete']('x');
  ^^^^^^^^

0 error(s), 2 warning(s), 100.0% typed
var a=new Map;a.set("x","x");a["delete"]("x");var b=new Set;b.add("x");b["delete"]("x");
```

I think `['delete']` is perfectly valid in this case and shouldn't throw any warnings.

**Comment by dimvar** [comment]
> I think ['delete'] is perfectly valid in this case and shouldn't throw any warnings.

Yes, it's valid JS, but the typechecker does not accept all valid JS. The error message is saying that you should use a dot when accessing the properties of a set.

**Comment by nazar-pc** [comment]
Well, it is generated by [LiveScript](http://livescript.net/) compiler, which I have no control over.

What is the fundamental issue with `Map` and `Set` that it doesn't allow `[]` there and can't be fixed?

**Comment by dimvar** [comment]
[They are marked as structs.](https://github.com/google/closure-compiler/blob/master/externs/es6_collections.js#L26)

Structs are more restrictive than general objects, and as a result they can be checked and optimized better by the compiler. For more, see [this](https://github.com/google/closure-compiler/wiki/%40struct-and-%40dict-Annotations).

**Comment by nazar-pc** [comment]
If the expression under `[]` is constant string, Closure Compiler should be able to understand `Map['delete']` as `Map.delete`. Don't you think this is a feasible improvement?

**Comment by MatrixFrog** [comment]
The LiveScript compiler is probably doing that because older versions of JS don't allow dotted property access for property names that are also keywords. I bet it would also write `promise['catch']` instead of `promise.catch`. If you're using Maps and Sets then you're not targeting those versions so just outputting `.delete` makes more sense. Do you think the LiveScript team would be open to changing their output? If not, you could probably write a simple `sed` replacement or something to do the conversion before passing the code on to Closure Compiler.
